File: frontend/apps/web/src/lib/features/flows/transcriptCorrections.ts
/**
 * Non-destructive transcript corrections, applied as a derived overlay on the
 * structured transcript lines (the same shape `applySpeakerNames` overlays).
 *
 * A correction is a char-range replacement anchored to the RAW segment text
 * the transcription step stored. The raw text is never rewritten, so every
 * corrected line can show what it was corrected from and be reverted.
 */

import type { TranscriptSegment } from "$lib/features/flows/transcriptSegments";

/** Wire shape of one stored correction (matches the backend API). */
export type CorrectionOccurrence = {
  segment_index: number;
  char_start: number;
  char_end: number;
  original: string;
  corrected: string;
};

export type AppliedCorrections = {
  segments: TranscriptSegment[];
  /** Raw (pre-correction) text per corrected segment index, for "corrected from". */
  correctedFrom: ReadonlyMap<number, string>;
  /** Occurrences whose anchor no longer matched; never applied approximately. */
  skipped: number;
};

/** Apply occurrences right-to-left per line so earlier offsets stay valid. */
export function applyTextCorrections(
  segments: readonly TranscriptSegment[],
  occurrences: readonly CorrectionOccurrence[]
): AppliedCorrections {
  if (occurrences.length === 0) {
    return { segments: [...segments], correctedFrom: new Map(), skipped: 0 };
  }
  const bySegment = new Map<number, CorrectionOccurrence[]>();
  let skipped = 0;
  for (const occurrence of occurrences) {
    if (occurrence.segment_index < 0 || occurrence.segment_index >= segments.length) {
      skipped += 1;
      continue;
    }
    const list = bySegment.get(occurrence.segment_index) ?? [];
    list.push(occurrence);
    bySegment.set(occurrence.segment_index, list);
  }
  const correctedFrom = new Map<number, string>();
  const result = segments.map((segment) => {
    const list = bySegment.get(segment.index);
    if (!list) return segment;
    let text = segment.text;
    let applied = false;
    const applicable = list.filter(
      (occurrence) =>
        occurrence.char_start >= 0 &&
        occurrence.char_end <= segment.text.length &&
        occurrence.char_start < occurrence.char_end &&
        segment.text.slice(occurrence.char_start, occurrence.char_end) === occurrence.original
    );
    skipped += list.length - applicable.length;
    for (const occurrence of [...applicable].sort((a, b) => b.char_start - a.char_start)) {
      text =
        text.slice(0, occurrence.char_start) +
        occurrence.corrected +
        text.slice(occurrence.char_end);
      applied = true;
    }
    if (!applied) return segment;
    correctedFrom.set(segment.index, segment.text);
    return { ...segment, text };
  });
  return { segments: result, correctedFrom, skipped };
}

export type LineEditDiff = {
  /** Minimal replacement in RAW-text space, or null when nothing changed. */
  occurrence: Omit<CorrectionOccurrence, "segment_index"> | null;
  /**
   * The changed region expanded to whole-token boundaries, when the change is
   * token-shaped (1-3 whole words replaced by non-empty text). This is what
   * seeds the "same correction elsewhere" search.
   */
  tokenShaped: { originalText: string; correctedText: string } | null;
};

const MAX_SUGGESTION_TOKENS = 3;

/** Minimal diff of one line edit: common prefix/suffix trim over the raw text. */
export function diffLineEdit(original: string, edited: string): LineEditDiff {
  if (original === edited) return { occurrence: null, tokenShaped: null };
  let start = 0;
  const maxStart = Math.min(original.length, edited.length);
  while (start < maxStart && original[start] === edited[start]) start += 1;
  let endOriginal = original.length;
  let endEdited = edited.length;
  while (
    endOriginal > start &&
    endEdited > start &&
    original[endOriginal - 1] === edited[endEdited - 1]
  ) {
    endOriginal -= 1;
    endEdited -= 1;
  }
  const occurrence = {
    char_start: start,
    char_end: endOriginal,
    original: original.slice(start, endOriginal),
    corrected: edited.slice(start, endEdited)
  };
  return { occurrence, tokenShaped: tokenShapedRegion(original, edited, occurrence) };
}

function tokenShapedRegion(
  original: string,
  edited: string,
  occurrence: Omit<CorrectionOccurrence, "segment_index">
): { originalText: string; correctedText: string } | null {
  const words = splitWords(original);
  // Closed-interval overlap so boundary contact counts: removing the space in
  // "Anna Lisa" must pull in both neighbouring tokens.
  const touched = words.filter(
    (word) =>
      word.offset <= occurrence.char_end && word.offset + word.word.length >= occurrence.char_start
  );
  if (touched.length === 0 || touched.length > MAX_SUGGESTION_TOKENS) return null;
  const lastTouched = touched[touched.length - 1];
  // Clamp so the region always contains the changed span; otherwise the
  // prefix/suffix mapping onto the edited string below would not hold.
  const regionStart = Math.min(touched[0].offset, occurrence.char_start);
  const regionEnd = Math.max(lastTouched.offset + lastTouched.word.length, occurrence.char_end);
  // The prefix [0, char_start) and suffix [char_end, …) are identical in both
  // strings, so expanding into them maps one-to-one onto the edited string.
  const shift = edited.length - original.length;
  const correctedText = edited.slice(regionStart, regionEnd + shift);
  if (correctedText.trim().length === 0) return null;
  if (/\s/.test(correctedText.trim()) && splitWords(correctedText).length > MAX_SUGGESTION_TOKENS) {
    return null;
  }
  return { originalText: original.slice(regionStart, regionEnd), correctedText };
}

export type WordSpan = { word: string; offset: number };

// A word is a run of letters/digits (Unicode-aware); surrounding punctuation
// stays out of the span so "sugary," matches the token "sugary". Never use
// ASCII \b here: names like "Çagri" have non-ASCII boundaries.
const WORD_RE = /[\p{L}\p{N}][\p{L}\p{N}'’-]*/gu;

export function splitWords(text: string): WordSpan[] {
  const words: WordSpan[] = [];
  for (const match of text.matchAll(WORD_RE)) {
    words.push({ word: match[0], offset: match.index ?? 0 });
  }
  return words;
}

function normalizeToken(token: string): string {
  return token.toLowerCase().normalize("NFD").replace(/\p{M}/gu, "");
}

/** Levenshtein distance with early exit above `limit`. */
export function boundedEditDistance(a: string, b: string, limit: number): number {
  if (Math.abs(a.length - b.length) > limit) return limit + 1;
  if (a === b) return 0;
  let previous = Array.from({ length: b.length + 1 }, (_, i) => i);
  for (let i = 1; i <= a.length; i += 1) {
    const current = [i];
    let rowMin = i;
    for (let j = 1; j <= b.length; j += 1) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      const value = Math.min(previous[j] + 1, current[j - 1] + 1, previous[j - 1] + cost);
      current.push(value);
      if (value < rowMin) rowMin = value;
    }
    if (rowMin > limit) return limit + 1;
    previous = current;
  }
  return previous[b.length];
}

export type OccurrenceCandidate = {
  segmentIndex: number;
  charStart: number;
  charEnd: number;
  matchedText: string;
  kind: "exact" | "fuzzy";
};

const FUZZY_DISTANCE_LIMIT = 2;
export const MAX_SUGGESTION_CANDIDATES = 50;

/**
 * Other places the corrected text likely appears: exact whole-token matches
 * (case-insensitive) plus tokens within edit distance 2 of the normalized
 * form, to catch different manglings of the same name. Multi-word originals
 * match joined consecutive tokens. Matching is whole-token only — a token
 * that merely contains the text ("Chagrin" for "Chagri") is matched as a
 * fuzzy candidate at best, never replaced inside the token.
 */
export function findOccurrences(
  segments: readonly TranscriptSegment[],
  originalText: string,
  exclude: { segmentIndex: number; charStart: number } | null
): OccurrenceCandidate[] {
  const targetWords = splitWords(originalText).map((span) => span.word);
  if (targetWords.length === 0) return [];
  const targetLower = targetWords.map((word) => word.toLowerCase());
  const targetNormalized = normalizeToken(targetWords.join(" "));
  const candidates: OccurrenceCandidate[] = [];
  for (const segment of segments) {
    const words = splitWords(segment.text);
    for (let i = 0; i + targetWords.length <= words.length; i += 1) {
      const window = words.slice(i, i + targetWords.length);
      const charStart = window[0].offset;
      const last = window[window.length - 1];
      const charEnd = last.offset + last.word.length;
      if (
        exclude &&
        segment.index === exclude.segmentIndex &&
        charStart <= exclude.charStart &&
        exclude.charStart < charEnd
      ) {
        continue;
      }
      const matchedText = segment.text.slice(charStart, charEnd);
      let kind: "exact" | "fuzzy" | null = null;
      if (window.every((span, j) => span.word.toLowerCase() === targetLower[j])) {
        kind = "exact";
      } else if (
        boundedEditDistance(
          normalizeToken(window.map((span) => span.word).join(" ")),
          targetNormalized,
          FUZZY_DISTANCE_LIMIT
        ) <= FUZZY_DISTANCE_LIMIT
      ) {
        kind = "fuzzy";
      }
      if (kind) {
        candidates.push({ segmentIndex: segment.index, charStart, charEnd, matchedText, kind });
        if (candidates.length >= MAX_SUGGESTION_CANDIDATES) return candidates;
      }
    }
  }
  return candidates;
}

/** Carry the matched text's casing over to the replacement ("SUGARY" → "ÇAGRI"). */
export function applyCasingPattern(matchedText: string, replacement: string): string {
  if (matchedText === matchedText.toUpperCase() && /\p{L}/u.test(matchedText)) {
    return replacement.toUpperCase();
  }
  const first = matchedText.match(/\p{L}/u)?.[0];
  if (first && first === first.toUpperCase()) {
    return replacement.charAt(0).toUpperCase() + replacement.slice(1);
  }
  return replacement;
}

/** Canonical order so saves and dirty checks are stable. */
export function sortOccurrences(
  occurrences: readonly CorrectionOccurrence[]
): CorrectionOccurrence[] {
  return [...occurrences].sort(
    (a, b) => a.segment_index - b.segment_index || a.char_start - b.char_start
  );
}

File: backend/src/eneo/flows/domain/transcript_corrections.py
"""Non-destructive transcript corrections for flow transcription steps.

A correction set stores char-range replacements anchored to the structured
transcript lines a transcription step persisted (``transcription.segments`` in
the step result's ``input_payload_json``). The raw transcript is never
rewritten: corrections are applied on read, so every corrected span keeps its
original text for auditing and revert.

Speaker edits live in the same set: a whole-segment edit reassigns a stored
line to another speaker label, a span edit reassigns part of a line. Both
anchor to the raw segment array exactly like text occurrences do.

Anchors are only valid for the exact segment array they were written against;
``segments_content_hash`` is the staleness token that detects re-transcription.
"""

from __future__ import annotations

import hashlib
import json
import re
from dataclasses import dataclass
from datetime import datetime
from typing import Any, cast
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from eneo.authentication.principal_types import PrincipalType
from eneo.flows.domain.speaker_labels import SPEAKER_LABEL_RE, SPEAKER_LINE_RE

TRANSCRIPT_CORRECTIONS_SCHEMA_VERSION = 2
SUPPORTED_TRANSCRIPT_CORRECTIONS_SCHEMA_VERSIONS = frozenset({1, 2})
MAX_CORRECTION_OCCURRENCES = 2000
MAX_SPEAKER_EDITS = 2000


@dataclass(frozen=True, slots=True)
class TranscriptCorrectionOccurrence:
    """One corrected span: ``original`` at ``[char_start, char_end)`` of the
    segment's text becomes ``corrected`` (empty string deletes the span)."""

    segment_index: int
    char_start: int
    char_end: int
    original: str
    corrected: str

    def as_json(self) -> dict[str, Any]:
        return {
            "segment_index": self.segment_index,
            "char_start": self.char_start,
            "char_end": self.char_end,
            "original": self.original,
            "corrected": self.corrected,
        }


@dataclass(frozen=True, slots=True)
class TranscriptSpeakerEdit:
    """Reassigns a segment, or a raw-text span of it, to another speaker.

    The null-span shape (``char_start``/``char_end``/``original`` all None)
    re-attributes the whole segment regardless of its text. A present span
    re-attributes exactly ``original`` at ``[char_start, char_end)`` of the
    raw text. ``original_speaker`` anchors both shapes to the segment's
    stored label, exactly as ``original`` anchors a text occurrence.
    """

    segment_index: int
    char_start: int | None
    char_end: int | None
    original: str | None
    original_speaker: str
    speaker: str

    @property
    def is_whole_segment(self) -> bool:
        return self.char_start is None

    def as_json(self) -> dict[str, Any]:
        return {
            "segment_index": self.segment_index,
            "char_start": self.char_start,
            "char_end": self.char_end,
            "original": self.original,
            "original_speaker": self.original_speaker,
            "speaker": self.speaker,
        }


@dataclass(eq=False)
class TranscriptCorrectionInvalidOccurrenceError(Exception):
    """A submitted occurrence does not fit the segment array it targets."""

    reason: str
    context: dict[str, Any]


@dataclass(eq=False)
class TranscriptSpeakerEditInvalidError(Exception):
    """A submitted speaker edit does not fit the segment array it targets."""

    reason: str
    context: dict[str, Any]


@dataclass(eq=False)
class FlowTranscriptCorrectionsStaleRevisionError(Exception):
    """The compare-and-swap revision no longer matches the stored row.

    ``current_revision`` is None when no row exists for the step;
    ``expected_revision`` is None when the caller tried to create a row that
    already exists.
    """

    expected_revision: int | None
    current_revision: int | None


class FlowTranscriptCorrectionSet(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    tenant_id: UUID
    flow_id: UUID
    flow_run_id: UUID
    step_id: UUID
    occurrences_json: list[dict[str, Any]]
    speaker_edits_json: list[dict[str, Any]] = Field(
        default_factory=list[dict[str, Any]]
    )
    revision: int
    schema_version: int
    segments_hash: str
    edited_by_user_id: UUID | None = None
    edited_by_service_id: UUID | None = None
    edited_by_principal_type: PrincipalType
    created_at: datetime
    updated_at: datetime

    def occurrences(self) -> list[TranscriptCorrectionOccurrence]:
        return [
            TranscriptCorrectionOccurrence(
                segment_index=int(item["segment_index"]),
                char_start=int(item["char_start"]),
                char_end=int(item["char_end"]),
                original=str(item["original"]),
                corrected=str(item["corrected"]),
            )
            for item in self.occurrences_json
        ]

    def speaker_edits(self) -> list[TranscriptSpeakerEdit]:
        return [
            TranscriptSpeakerEdit(
                segment_index=int(item["segment_index"]),
                char_start=(
                    int(item["char_start"]) if item["char_start"] is not None else None
                ),
                char_end=(
                    int(item["char_end"]) if item["char_end"] is not None else None
                ),
                original=(
                    str(item["original"]) if item["original"] is not None else None
                ),
                original_speaker=str(item["original_speaker"]),
                speaker=str(item["speaker"]),
            )
            for item in self.speaker_edits_json
        ]


def segments_content_hash(raw_segments: list[dict[str, Any]]) -> str:
    """Staleness token for a stored segment array.

    Canonical JSON so key order and whitespace never change the hash; any
    re-transcription that changes a single character invalidates anchors.
    """
    canonical = json.dumps(
        raw_segments,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
    )
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()


def sort_occurrences(
    occurrences: list[TranscriptCorrectionOccurrence],
) -> list[TranscriptCorrectionOccurrence]:
    """Canonical order so saves, dirty checks, and diffs are stable."""
    return sorted(occurrences, key=lambda item: (item.segment_index, item.char_start))


def sort_speaker_edits(
    speaker_edits: list[TranscriptSpeakerEdit],
) -> list[TranscriptSpeakerEdit]:
    """Canonical order: whole-segment edits first, then spans by offset."""
    return sorted(
        speaker_edits,
        key=lambda item: (
            item.segment_index,
            -1 if item.char_start is None else item.char_start,
        ),
    )


def _segment_text(segment: dict[str, Any]) -> str:
    text = segment.get("text")
    return text if isinstance(text, str) else ""


def _segment_speaker(segment: dict[str, Any]) -> str | None:
    speaker = segment.get("speaker")
    return speaker if isinstance(speaker, str) and speaker else None


def validate_occurrences(
    segments: list[dict[str, Any]],
    occurrences: list[TranscriptCorrectionOccurrence],
) -> None:
    """Reject occurrence lists that do not fit ``segments`` exactly.

    Raises TranscriptCorrectionInvalidOccurrenceError for the first violation:
    out-of-range segment index, invalid char range, a span whose ``original``
    no longer matches the anchored text, overlapping ranges within a segment,
    or an oversized list.
    """
    if len(occurrences) > MAX_CORRECTION_OCCURRENCES:
        raise TranscriptCorrectionInvalidOccurrenceError(
            reason="too_many_occurrences",
            context={
                "occurrence_count": len(occurrences),
                "max_occurrences": MAX_CORRECTION_OCCURRENCES,
            },
        )
    by_segment: dict[int, list[TranscriptCorrectionOccurrence]] = {}
    for occurrence in occurrences:
        if not 0 <= occurrence.segment_index < len(segments):
            raise TranscriptCorrectionInvalidOccurrenceError(
                reason="segment_index_out_of_range",
                context={
                    "segment_index": occurrence.segment_index,
                    "segment_count": len(segments),
                },
            )
        text = _segment_text(segments[occurrence.segment_index])
        if not 0 <= occurrence.char_start < occurrence.char_end <= len(text):
            raise TranscriptCorrectionInvalidOccurrenceError(
                reason="char_range_invalid",
                context={
                    "segment_index": occurrence.segment_index,
                    "char_start": occurrence.char_start,
                    "char_end": occurrence.char_end,
                    "text_length": len(text),
                },
            )
        anchored = text[occurrence.char_start : occurrence.char_end]
        if anchored != occurrence.original:
            raise TranscriptCorrectionInvalidOccurrenceError(
                reason="original_mismatch",
                context={
                    "segment_index": occurrence.segment_index,
                    "char_start": occurrence.char_start,
                    "char_end": occurrence.char_end,
                    "anchored_text": anchored,
                    "original": occurrence.original,
                },
            )
        by_segment.setdefault(occurrence.segment_index, []).append(occurrence)
    for segment_index, segment_occurrences in by_segment.items():
        ordered = sorted(segment_occurrences, key=lambda item: item.char_start)
        for previous, current in zip(ordered, ordered[1:]):
            if current.char_start < previous.char_end:
                raise TranscriptCorrectionInvalidOccurrenceError(
                    reason="overlapping_ranges",
                    context={
                        "segment_index": segment_index,
                        "first_range": [previous.char_start, previous.char_end],
                        "second_range": [current.char_start, current.char_end],
                    },
                )


def validate_speaker_edits(
    segments: list[dict[str, Any]],
    speaker_edits: list[TranscriptSpeakerEdit],
) -> None:
    """Reject speaker-edit lists that do not fit ``segments`` exactly.

    Raises TranscriptSpeakerEditInvalidError for the first violation. A
    whole-segment edit is exclusive with any other edit on the same segment;
    span edits must not overlap each other (adjacent spans are fine). An edit
    that keeps the stored speaker is rejected: revert is expressed by removing
    the edit, never by writing a no-op.
    """
    if len(speaker_edits) > MAX_SPEAKER_EDITS:
        raise TranscriptSpeakerEditInvalidError(
            reason="too_many_speaker_edits",
            context={
                "speaker_edit_count": len(speaker_edits),
                "max_speaker_edits": MAX_SPEAKER_EDITS,
            },
        )
    by_segment: dict[int, list[TranscriptSpeakerEdit]] = {}
    for edit in speaker_edits:
        if not 0 <= edit.segment_index < len(segments):
            raise TranscriptSpeakerEditInvalidError(
                reason="segment_index_out_of_range",
                context={
                    "segment_index": edit.segment_index,
                    "segment_count": len(segments),
                },
            )
        for label in (edit.original_speaker, edit.speaker):
            if not SPEAKER_LABEL_RE.match(label):
                raise TranscriptSpeakerEditInvalidError(
                    reason="invalid_speaker_label",
                    context={"segment_index": edit.segment_index, "label": label},
                )
        segment = segments[edit.segment_index]
        stored_speaker = _segment_speaker(segment)
        if stored_speaker is None:
            raise TranscriptSpeakerEditInvalidError(
                reason="segment_has_no_speaker",
                context={"segment_index": edit.segment_index},
            )
        if edit.original_speaker != stored_speaker:
            raise TranscriptSpeakerEditInvalidError(
                reason="original_speaker_mismatch",
                context={
                    "segment_index": edit.segment_index,
                    "original_speaker": edit.original_speaker,
                    "stored_speaker": stored_speaker,
                },
            )
        if edit.char_start is None or edit.char_end is None or edit.original is None:
            all_null = (
                edit.char_start is None
                and edit.char_end is None
                and edit.original is None
            )
            if not all_null:
                raise TranscriptSpeakerEditInvalidError(
                    reason="char_range_invalid",
                    context={
                        "segment_index": edit.segment_index,
                        "char_start": edit.char_start,
                        "char_end": edit.char_end,
                    },
                )
        else:
            text = _segment_text(segment)
            if not 0 <= edit.char_start < edit.char_end <= len(text):
                raise TranscriptSpeakerEditInvalidError(
                    reason="char_range_invalid",
                    context={
                        "segment_index": edit.segment_index,
                        "char_start": edit.char_start,
                        "char_end": edit.char_end,
                        "text_length": len(text),
                    },
                )
            anchored = text[edit.char_start : edit.char_end]
            if anchored != edit.original:
                raise TranscriptSpeakerEditInvalidError(
                    reason="original_mismatch",
                    context={
                        "segment_index": edit.segment_index,
                        "char_start": edit.char_start,
                        "char_end": edit.char_end,
                        "anchored_text": anchored,
                        "original": edit.original,
                    },
                )
        if edit.speaker == stored_speaker:
            raise TranscriptSpeakerEditInvalidError(
                reason="speaker_unchanged",
                context={
                    "segment_index": edit.segment_index,
                    "speaker": edit.speaker,
                },
            )
        by_segment.setdefault(edit.segment_index, []).append(edit)
    for segment_index, segment_edits in by_segment.items():
        wholes = [edit for edit in segment_edits if edit.is_whole_segment]
        spans = [edit for edit in segment_edits if not edit.is_whole_segment]
        if wholes and (spans or len(wholes) > 1):
            raise TranscriptSpeakerEditInvalidError(
                reason="whole_segment_conflicts_with_span",
                context={"segment_index": segment_index},
            )
        ordered = sorted((edit.char_start or 0, edit.char_end or 0) for edit in spans)
        for previous, current in zip(ordered, ordered[1:]):
            if current[0] < previous[1]:
                raise TranscriptSpeakerEditInvalidError(
                    reason="overlapping_speaker_spans",
                    context={
                        "segment_index": segment_index,
                        "first_range": [previous[0], previous[1]],
                        "second_range": [current[0], current[1]],
                    },
                )


def _applicable_text_occurrences(
    text: str,
    occurrences: list[TranscriptCorrectionOccurrence],
    skipped: list[TranscriptCorrectionOccurrence],
) -> list[TranscriptCorrectionOccurrence]:
    """Anchor-verified occurrences for one segment, sorted by offset."""
    applicable: list[TranscriptCorrectionOccurrence] = []
    for occurrence in occurrences:
        in_range = 0 <= occurrence.char_start < occurrence.char_end <= len(text)
        if (
            not in_range
            or text[occurrence.char_start : occurrence.char_end] != occurrence.original
        ):
            skipped.append(occurrence)
            continue
        applicable.append(occurrence)
    return sorted(applicable, key=lambda item: item.char_start)


def _map_raw_offset(
    boundary: int,
    applicable: list[TranscriptCorrectionOccurrence],
) -> int:
    """Map a raw-text offset into corrected-text space (monotone, total).

    ``applicable`` must be this segment's anchor-verified occurrences sorted
    ascending. A boundary strictly inside a replaced span clamps inside the
    replacement so runs stay contiguous whatever the replacement length.
    """
    delta = 0
    for occurrence in applicable:
        if occurrence.char_end <= boundary:
            delta += len(occurrence.corrected) - (
                occurrence.char_end - occurrence.char_start
            )
        elif occurrence.char_start < boundary:
            offset_in_span = min(
                boundary - occurrence.char_start, len(occurrence.corrected)
            )
            return occurrence.char_start + delta + offset_in_span
        else:
            break
    return boundary + delta


def _corrected_speaker_runs(
    raw_text: str,
    stored_speaker: str,
    spans: list[TranscriptSpeakerEdit],
    applicable: list[TranscriptCorrectionOccurrence],
    skipped: list[TranscriptSpeakerEdit],
) -> list[dict[str, Any]]:
    """Consecutive speaker runs over the corrected text, merged and non-empty.

    ``spans`` are this segment's anchor-verified span edits. A single run in
    the stored speaker means nothing changed and the result is empty.
    """
    raw_runs: list[tuple[int, int, str]] = []
    cursor = 0
    for edit in sorted(spans, key=lambda item: item.char_start or 0):
        start = edit.char_start or 0
        end = edit.char_end or 0
        if start < cursor:
            # Overlap can only mean corrupt storage; skip, never guess.
            skipped.append(edit)
            continue
        if cursor < start:
            raw_runs.append((cursor, start, stored_speaker))
        raw_runs.append((start, end, edit.speaker))
        cursor = end
    if cursor < len(raw_text):
        raw_runs.append((cursor, len(raw_text), stored_speaker))

    merged: list[dict[str, Any]] = []
    for raw_start, raw_end, speaker in raw_runs:
        start = _map_raw_offset(raw_start, applicable)
        end = _map_raw_offset(raw_end, applicable)
        if start >= end:
            continue
        if merged and merged[-1]["speaker"] == speaker:
            merged[-1]["char_end"] = end
            continue
        merged.append({"char_start": start, "char_end": end, "speaker": speaker})
    if len(merged) == 1 and merged[0]["speaker"] == stored_speaker:
        return []
    return merged


def apply_corrections_and_speaker_edits(
    segments: list[dict[str, Any]],
    occurrences: list[TranscriptCorrectionOccurrence],
    speaker_edits: list[TranscriptSpeakerEdit],
) -> tuple[
    list[dict[str, Any]],
    list[TranscriptCorrectionOccurrence],
    list[TranscriptSpeakerEdit],
]:
    """Corrected copies of ``segments`` plus whatever could not be applied.

    Text corrections replace within ``text`` right-to-left. Speaker edits
    then attribute content: a whole-segment edit replaces ``speaker``; span
    edits attach ``speaker_runs`` (consecutive ``{char_start, char_end,
    speaker}`` runs covering the corrected text, raw boundaries mapped through
    the applied corrections) unless they collapse to a single speaker, which
    replaces ``speaker`` instead. Mismatched anchors are skipped, never
    applied approximately.
    """
    skipped_occurrences: list[TranscriptCorrectionOccurrence] = []
    skipped_speaker_edits: list[TranscriptSpeakerEdit] = []
    occurrences_by_segment: dict[int, list[TranscriptCorrectionOccurrence]] = {}
    for occurrence in occurrences:
        if not 0 <= occurrence.segment_index < len(segments):
            skipped_occurrences.append(occurrence)
            continue
        occurrences_by_segment.setdefault(occurrence.segment_index, []).append(
            occurrence
        )
    edits_by_segment: dict[int, list[TranscriptSpeakerEdit]] = {}
    for edit in speaker_edits:
        if not 0 <= edit.segment_index < len(segments):
            skipped_speaker_edits.append(edit)
            continue
        edits_by_segment.setdefault(edit.segment_index, []).append(edit)

    corrected_segments = [dict(segment) for segment in segments]
    touched = sorted(set(occurrences_by_segment) | set(edits_by_segment))
    for segment_index in touched:
        raw_text = _segment_text(segments[segment_index])
        stored_speaker = _segment_speaker(segments[segment_index])
        applicable = _applicable_text_occurrences(
            raw_text,
            occurrences_by_segment.get(segment_index, []),
            skipped_occurrences,
        )
        text = raw_text
        for occurrence in reversed(applicable):
            text = (
                text[: occurrence.char_start]
                + occurrence.corrected
                + text[occurrence.char_end :]
            )
        if segment_index in occurrences_by_segment:
            corrected_segments[segment_index]["text"] = text

        applicable_edits: list[TranscriptSpeakerEdit] = []
        for edit in sort_speaker_edits(edits_by_segment.get(segment_index, [])):
            anchored = (
                stored_speaker is not None and edit.original_speaker == stored_speaker
            )
            if anchored and not edit.is_whole_segment:
                anchored = (
                    edit.char_start is not None
                    and edit.char_end is not None
                    and 0 <= edit.char_start < edit.char_end <= len(raw_text)
                    and raw_text[edit.char_start : edit.char_end] == edit.original
                )
            if not anchored:
                skipped_speaker_edits.append(edit)
                continue
            applicable_edits.append(edit)
        wholes = [edit for edit in applicable_edits if edit.is_whole_segment]
        spans = [edit for edit in applicable_edits if not edit.is_whole_segment]
        if wholes:
            # Canonical sets hold at most one whole-segment edit and no
            # coexisting spans; anything extra is corrupt storage — skip it.
            corrected_segments[segment_index]["speaker"] = wholes[0].speaker
            skipped_speaker_edits.extend(wholes[1:])
            skipped_speaker_edits.extend(spans)
        elif spans and stored_speaker is not None:
            runs = _corrected_speaker_runs(
                raw_text, stored_speaker, spans, applicable, skipped_speaker_edits
            )
            if len(runs) == 1:
                corrected_segments[segment_index]["speaker"] = runs[0]["speaker"]
            elif runs:
                corrected_segments[segment_index]["speaker_runs"] = runs
    return corrected_segments, skipped_occurrences, skipped_speaker_edits


def apply_corrections(
    segments: list[dict[str, Any]],
    occurrences: list[TranscriptCorrectionOccurrence],
) -> tuple[list[dict[str, Any]], list[TranscriptCorrectionOccurrence]]:
    """Return corrected copies of ``segments`` plus the skipped occurrences.

    Occurrences whose anchor no longer matches (or no longer exists) are
    skipped, never applied approximately. Within a segment, replacements run
    right-to-left so earlier char offsets stay valid.
    """
    corrected_segments, skipped, _ = apply_corrections_and_speaker_edits(
        segments, occurrences, []
    )
    return corrected_segments, skipped


def apply_to_rendered_transcript(
    rendered_text: str,
    segments: list[dict[str, Any]],
    occurrences: list[TranscriptCorrectionOccurrence],
    speaker_edits: list[TranscriptSpeakerEdit],
) -> str | None:
    """Patch the rendered transcript lines, or None when they do not align.

    The rendered text is patchable only while its speaker lines correspond
    1:1, in order, to the stored segments (same label, same text). Any
    mismatch — a hand-edited line, a drifted render — returns None; callers
    must skip propagation rather than partial-apply. Non-speaker lines (part
    headers, blanks) pass through untouched. A segment attributed to several
    speakers renders one line per run, reusing the segment's timestamp prefix.
    """
    lines = rendered_text.split("\n")
    matches: list[tuple[int, re.Match[str]]] = []
    for line_index, line in enumerate(lines):
        match = SPEAKER_LINE_RE.match(line)
        if match:
            matches.append((line_index, match))
    if len(matches) != len(segments):
        return None
    for (_, match), segment in zip(matches, segments):
        if match.group("label") != _segment_speaker(segment):
            return None
        if match.group("text") != _segment_text(segment):
            return None
    corrected_segments, _, _ = apply_corrections_and_speaker_edits(
        segments, occurrences, speaker_edits
    )
    for (line_index, match), corrected in zip(matches, corrected_segments):
        prefix = match.group("prefix")
        text = _segment_text(corrected)
        speaker = _segment_speaker(corrected)
        runs = corrected.get("speaker_runs")
        rendered_runs: list[str] = []
        if isinstance(runs, list):
            for run in cast("list[dict[str, Any]]", runs):
                run_text = text[run["char_start"] : run["char_end"]].strip()
                if run_text:
                    rendered_runs.append(f"{prefix}{run['speaker']}: {run_text}")
        if rendered_runs:
            lines[line_index] = "\n".join(rendered_runs)
        else:
            lines[line_index] = f"{prefix}{speaker}: {text}"
    return "\n".join(lines)
